datafusion/src/optimizer/simplify_expressions.rs

```diff
@@ -662,6 +664,48 @@ impl<'a> ExprRewriter for Simplifier<'a> {
                 _ => unreachable!(),
             },
 
+            //
+            // Rules for Case
+            //
+
+            // CASE
+            //   WHEN X THEN A
+            //   WHEN Y THEN B
+            //   ...
+            //   ELSE Q
+            // END
+            //
+            // ---> (X AND A) OR (Y AND B AND NOT X) OR ... (NOT (X OR Y) AND Q)
+            Case {
+                expr: None,
+                when_then_expr,
+                else_expr,
+            } if is_boolean_expr => {
+                // The disjunction of all the when predicates encountered so far
```

Comment:
Given the complexity of expression this could potentially produce `O(n!)` maybe in terms of the number of `when` exprs, what do you think about adding a heuristic that only does this rewrite when there are a 'small' number of cases -- perhaps 2 or 3?

This would prevent some sort of pathological explosion and would handle the usecase of simple mapping `CASE` statements